frontend/app/common/fetcher.ts

```diff
@@ -1,100 +1,73 @@
-const methods: FetcherMethod[] = ['get', 'post', 'put', 'patch', 'delete', 'head'];
-
-interface FetcherInitBase {
-  url: string;
-  overriddenApiBase?: string;
-  withCredentials?: boolean;
-  /** whether log error message to console */
-  logError?: boolean;
-}
-
-interface FetcherInitJSON extends FetcherInitBase {
-  contentType?: 'application/json';
-  body?: string | object | Blob | ArrayBuffer;
-}
-
-interface FetcherInitMultipart extends FetcherInitBase {
-  contentType: 'multipart/form-data';
-  body: FormData;
-}
-
-type FetcherInit = string | FetcherInitJSON | FetcherInitMultipart;
-
-type FetcherObject = { [K in FetcherMethod]: <T = unknown>(data: FetcherInit) => Promise<T> };
+import { siteId } from 'common/settings';
+
+/** List of fetcher’s supported methods */
+const METHODS = ['get', 'post', 'put', 'delete'] as const;
+/** Header name for JWT token */
+export const JWT_HEADER = 'X-JWT';
+/** Header name for XSRF token */
+export const XSRF_HEADER = 'X-XSRF-TOKEN';
+/** Cookie field with XSRF token */
+export const XSRF_COOKIE = 'XSRF-TOKEN';
+
+type QueryParams = Record<string, string | number | undefined>;
+type Payload = BodyInit | Record<string, unknown> | null;
+type FetcherMethods = Record<'get' | 'delete', <T>(url: string, query?: QueryParams) => Promise<T>> &
+  Record<'put' | 'post', <T>(url: string, query?: QueryParams, body?: Payload) => Promise<T>>;
 
 /** JWT token received from server and will be send by each request, if it present */
 let activeJwtToken: string | undefined;
 
-const fetcher = methods.reduce<Partial<FetcherObject>>((acc, method) => {
-  acc[method] = async <T = unknown>(data: FetcherInit): Promise<T> => {
-    const {
-      url,
-      body = undefined,
-      withCredentials = false,
-      overriddenApiBase = API_BASE,
-      contentType = 'application/json',
-      logError = true,
-    } = typeof data === 'string' ? { url: data } : data;
-    const baseUrl = `${BASE_URL}${overriddenApiBase}`;
-
-    const headers = new Headers({
-      Accept: 'application/json',
-      'X-XSRF-TOKEN': getCookie('XSRF-TOKEN') || '',
-    });
-
-    if (contentType !== 'multipart/form-data') {
-      headers.append('Content-Type', contentType);
-    }
-
-    // Save token in memory and pass it into headers in case if storing cookies is disabled
-    if (activeJwtToken) {
-      headers.append(HEADER_X_JWT, activeJwtToken);
-    }
-
-    let rurl = `${baseUrl}${url}`;
-
-    const parameters: RequestInit = {
-      method,
-      headers,
-      mode: 'cors',
-      credentials: withCredentials ? 'include' : 'omit',
-    };
-
-    if (body) {
-      if (contentType === 'multipart/form-data') {
-        parameters.body = body as FormData;
-      } else if (typeof body === 'object' && !(body instanceof Blob) && !(body instanceof ArrayBuffer)) {
-        parameters.body = JSON.stringify(body);
+const createFetcher = (baseUrl: string = '') =>
+  METHODS.reduce<FetcherMethods>((acc, method) => {
```

Comment:
It is looks little messy. I think will be better to define each method separate.